nvme.c

```diff
@@ -8858,16 +8871,32 @@ static int gen_tls_key(int argc, char **argv, struct command *command, struct pl
 		}
 	}
 
-	crc = crc32(crc, raw_secret, key_len);
-	raw_secret[key_len++] = crc & 0xff;
-	raw_secret[key_len++] = (crc >> 8) & 0xff;
-	raw_secret[key_len++] = (crc >> 16) & 0xff;
-	raw_secret[key_len++] = (crc >> 24) & 0xff;
+	if (!cfg.hostnqn && !cfg.subsysnqn) {
+		crc = crc32(crc, raw_secret, key_len);
+		raw_secret[key_len++] = crc & 0xff;
+		raw_secret[key_len++] = (crc >> 8) & 0xff;
+		raw_secret[key_len++] = (crc >> 16) & 0xff;
+		raw_secret[key_len++] = (crc >> 24) & 0xff;
 
-	memset(encoded_key, 0, sizeof(encoded_key));
-	base64_encode(raw_secret, key_len, encoded_key);
+		memset(encoded_key, 0, sizeof(encoded_key));
+		base64_encode(raw_secret, key_len, encoded_key);
+		printf("NVMeTLSkey-1:%02x:%s:\n",
+		       cfg.hmac, encoded_key);
+		return 0;
```

Comment:
I don't understand the semantics of this function. if host and subsys nqns not given you just print the key and exit, and if they were given then it is stored in a keyring?

Reply:
Correct. The 'NVMeTLSkey' thingie is the so-called 'PSK Interchange Format' (cf section 3.6.1.5 of the NVMe TCP spec), and is meant to be used if you want to generate a PSK to be installed on (other) machines.
This format has _no_ indication to which NQN is should match.
From this the actual TLS key should be derived (cf section 3.6.1.3 of the NVMe TCP spec), for which you have to specify the host NQN and subsysten NQN as they are factored in when deriving the TLS key.